py/server/tests/test_vectorization.py

```diff
@@ -0,0 +1,202 @@
+
+        seed = 10
+        t = empty_table(1).update("X = py_const(seed)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        t = empty_table(1).update("X = py_const(30*1024*1034*1034)")
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        t = empty_table(1).update("X = py_const(30000000000L)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        t = empty_table(1).update("X = py_const(100.01)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        t = empty_table(1).update("X = py_const(100.01f)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(1).update("X = py_const(NULL_INT)")
+        self.assertIn("NULL_INT", str(cm.exception))
+
+        def py_const_str(s) -> str:
+            return "hello " + str(s) + "!"
+
+        t = empty_table(1).update("X = py_const_str(`Deephaven`)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(1).update("X = py_const_str(null)")
+        expected_count += 1
+        self.assertIn("non null", str(cm.exception))
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+        t = empty_table(1).update("X = py_const_str(true)")
+        expected_count += 1
+        self.assertEqual(deephaven.table._vectorized_count, expected_count)
+
+    def test_multiple_formulas(self):
+        def pyfunc(p1, p2, p3) -> int:
+            return p1 + p2 + p3
+
+        t = empty_table(1).update("X = i").update(["Y = pyfunc(X, i, 33)", "Z = pyfunc(X, ii, 66)"])
+        self.assertEqual(deephaven.table._vectorized_count, 2)
+        self.assertIn("33", t.to_string(cols=["Y"]))
+        self.assertIn("66", t.to_string(cols=["Z"]))
+
+    def test_multiple_formulas_vectorized(self):
+        @DhVectorize
+        def pyfunc(p1, p2, p3) -> int:
+            return p1 + p2 + p3
+
+        t = empty_table(1).update("X = i").update(["Y = pyfunc(X, i, 33)", "Z = pyfunc(X, ii, 66)"])
+        self.assertEqual(deephaven.table._vectorized_count, 3)
+        self.assertIn("33", t.to_string(cols=["Y"]))
+        self.assertIn("66", t.to_string(cols=["Z"]))
+
+    def test_filters(self):
+        def pyfunc_int(p1, p2, p3) -> int:
+            return p1 * p2 * p3
+
+        def pyfunc_bool(p1, p2, p3) -> bool:
+            return p1 * p2 * p3
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(10).view(formulas=["I=ii", "J=(ii * 2)"]).where("pyfunc_int(I, 3, J)")
+        self.assertEqual(deephaven.table._vectorized_count, 0)
+        self.assertIn("boolean required", str(cm.exception))
```

Comment:
Does `pyfunc_int(I,3,J) == 1` work?  Either way, it should be a test case.  I think it doesn't.

Reply:
It doesn't.